### PR TITLE
cranelift: only allow vector types with saturating arithmetic

### LINKED ISSUE #2883: Cranelift: `unreachable` in `cranelift-codegen/src/isa/x64/lower.rs:1800:26` for `Opcode::SsubSat`
### `.clif` Test Case

```
function u0:0() -> i64 system_v {
    block0:
        v0 = iconst.i64 1
        v1 = iconst.i64 1
        v2 = ssub_sat v0, v1
        return v2
}
```

### Steps to Reproduce

Use `ssub_sat` with two `I64` values and (at least for my case) use `cranelift-object` with a triple that looks like:

```
target_lexicon::Triple::host() = Triple {
    architecture: X86_64,
    vendor: Unknown,
    operating_system: Linux,
    environment: Gnu,
    binary_format: Elf,
}
```

### Expected Results

The code should compile correctly and the function should return `0`

### Actual Results

```
thread 'main' panicked at 'internal error: entered unreachable code: SsubSat', /home/mental/.cargo/registry/src/github.com-1ecc6299db9ec823/cranelift-codegen-0.73.0/src/isa/x64/lower.rs:1800:26
```

(note that the unreachable message is actually empty in the code, I have modified it locally to emit the unsupported opcode)

an unreachable case is hit instead.

### Versions and Environment

Cranelift version or commit: `0.73.0`

Operating system: `Linux felix 5.11.15-arch1-2 #1 SMP PREEMPT Sat, 17 Apr 2021 00:22:30 +0000 x86_64 GNU/Linux`

Architecture: `x86_64`

### Extra Info

<details>
  <summary>minimal, complete and verifiable example (mcve)</summary>
  
```rust
use cranelift_codegen::{binemit, ir::InstBuilder, Context};
use cranelift_codegen::{
    ir::{self, Signature},
    isa, settings, verify_function,
};
use cranelift_frontend::{FunctionBuilder, FunctionBuilderContext};
use cranelift_module::{Linkage, Module};
use cranelift_object::{ObjectBuilder, ObjectModule};
use settings::FlagsOrIsa;

fn main() {
    let target_isa = isa::lookup(dbg!(target_lexicon::Triple::host()))
        .unwrap()
        .finish(settings::Flags::new(settings::builder()));

    let object_builder = ObjectBuilder::new(
        target_isa,
        "<empty>".to_string(),
        cranelift_module::default_libcall_names(),
    )
    .unwrap();

    let target_isa = isa::lookup(target_lexicon::Triple::host())
        .unwrap()
        .finish(settings::Flags::new(settings::builder()));

    let mut object_module = ObjectModule::new(object_builder);

    let mut sig = Signature::new(cranelift_codegen::isa::CallConv::SystemV);

    sig.returns.push(ir::AbiParam::new(ir::types::I64));

    let mut func = ir::Function::with_name_signature(
        ir::ExternalName::User {
            namespace: 0,
            index: 0,
        },
        sig,
    );

    let mut builder_ctx = FunctionBuilderContext::new();
    let mut builder = FunctionBuilder::new(&mut func, &mut builder_ctx);

    let b0 = builder.create_block();

    builder.switch_to_block(b0);

    let v1 = builder.ins().iconst(ir::types::I64, 1);
    let v2 = builder.ins().iconst(ir::types::I64, 1);

    let v3 = builder.ins().ssub_sat(v1, v2);

    builder.ins().return_(&[v3]);

    let fisa = FlagsOrIsa::from(&*target_isa);

    verify_function(&builder.func, fisa).unwrap();

    let fid = object_module
        .declare_function(&"main", Linkage::Export, &func.signature.clone())
        .unwrap();

    let mut ctx = Context::for_function(func);

    object_module
        .define_function(
            fid,
            &mut ctx,
            &mut binemit::NullTrapSink {},
            &mut binemit::NullStackMapSink {},
        )
        .unwrap();
}
```

```toml
[package]
name = "clif-usub"
version = "0.1.0"
edition = "2018"

[dependencies]
cranelift-codegen = "0.73.0"
cranelift-frontend = "0.73.0"
cranelift-object = "0.73.0"
cranelift-module = "0.73.0"
target-lexicon = "0.12.0"
```

</details>

### PR DESCRIPTION
**Pull request by abrown**
This fixes #2883 by restricting which types are available to the `uadd_sat`, `sadd_sat`, `usub_sat`, and `ssub_sat` IR operations.

### CODE REVIEW

**Comment by abrown** [comment]
There are other considerations to this:
 - perhaps the arm32 and aarch64 backends could be simplified now that they don't have to lower the scalar versions of these operations
 - we could alternately implement the operations for x64 but I don't particularly see the need at the moment: the lowering is not optimal and these operations were introduced to implement the Wasm SIMD specification.

**Comment by bjorn3** [comment]
They could be useful to lower the saturating arithmetic intrinsics of rust to.

**Review by bnjbvr — CHANGES_REQUESTED** [review]
That makes sense to me that the IR builder would make sure to use the right opcodes and do the conversions from scalar to vec (and conversely) by themselves. cc @cfallin for visibility into this change.

If we want to go down this road, can you also get rid of the scalar handling in aarch64's lowering of all these opcodes please? Otherwise this is technically unreachable code. Maybe we could replace the `if` by an assert that the input types are vector types.

**Comment by cfallin** [comment]
+1 to @bnjbvr's suggestion -- the lowering code for these instructions in the ARM backends should go away. I think that when I wrote it for aarch64 I just wasn't aware of the vector-only history; cool that it was supported but less code is always better from a maintenance perspective, especially if it is now dead code!

File: cranelift/codegen/src/isa/aarch64/lower_inst.rs [hunk 1]
```diff
@@ -150,68 +150,27 @@ pub(crate) fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
         Opcode::UaddSat | Opcode::SaddSat | Opcode::UsubSat | Opcode::SsubSat => {
             // We use the scalar SIMD & FP saturating additions and subtractions
             // (SQADD / UQADD / SQSUB / UQSUB), which require scalar FP registers.
```
Comment (abrown) [inline cranelift/codegen/src/isa/aarch64/lower_inst.rs:152]: I think I got the below re-factoring correct but this comment may need to change--suggestions?

**Comment by abrown** [comment]
@fitzgen, I am assuming peepmatic had to be updated because of the cranelift/meta changes but correct me if I did something wrong: I simply ran `cargo test --features 'enable-peepmatic rebuild-peephole-optimizers' peepmatic` and committed the results.

**Comment by fitzgen** [comment]
> @fitzgen, I am assuming peepmatic had to be updated because of the cranelift/meta changes but correct me if I did something wrong: I simply ran `cargo test --features 'enable-peepmatic rebuild-peephole-optimizers' peepmatic` and committed the results.

That sounds correct :+1:

File: cranelift/codegen/src/isa/aarch64/lower_inst.rs (hunk 1 above)
Comment (akirilov-arm) [reply cranelift/codegen/src/isa/aarch64/lower_inst.rs:152]: Just remove the comment completely - it's no longer relevant.

Comment (abrown) [reply cranelift/codegen/src/isa/aarch64/lower_inst.rs:152]: Thanks!

**Review by akirilov-arm — APPROVED** [review]
The AArch64-related changes look good to me, but I'd wait for @bnjbvr to approve too (since he had feedback before).

**Review by bnjbvr — APPROVED** [review]
Thanks!

**Pull request merged**